openfisca_core/tracers.py

```diff
@@ -130,53 +162,125 @@ def _get_time_in_sec(self) -> float:
+            ]
+
+        self._write_csv(os.path.join(dir_path, 'aggregated_performance_table.csv'), aggregated_csv_rows)
+
+    def generate_performance_details(self, dir_path: str) -> None:
+        self.generate_performance_graph(dir_path)
+        self.generate_performance_tables(dir_path)
+
+    def _write_csv(self, path: str, rows: List[Dict[str, Any]]) -> None:
+        fieldnames = list(rows[0].keys())
+        with open(path, 'w') as csv_file:
+            writer = csv.DictWriter(csv_file, fieldnames = fieldnames)
+            writer.writeheader()
+            for row in rows:
+                writer.writerow(row)
+
+    def aggregate_calculation_times(self, flat_trace: Dict) -> Dict[str, Dict]:
+
+        def _aggregate_calculations(calculations):
+            nof_calculations = len(calculations)
+            calculation_time = sum(calculation[1]['calculation_time'] for calculation in calculations)
+            formula_time = sum(calculation[1]['formula_time'] for calculation in calculations)
+            return {
+                'nof_calculations': nof_calculations,
+                'calculation_time': calculation_time,
+                'formula_time': formula_time,
+                'avg_calculation_time': calculation_time / nof_calculations,
+                'avg_formula_time': formula_time / nof_calculations
+                }
+
+        all_calculations = sorted(flat_trace.items())
+        return {
+            variable_name: _aggregate_calculations(list(calculations))
+            for variable_name, calculations in groupby(all_calculations, lambda calculation: calculation[0].split('<')[0])
+            }
+
     def _get_nb_requests(self, tree, variable: str):
-        tree_call = tree['name'] == variable
-        children_calls = sum(self._get_nb_requests(child, variable) for child in tree['children'])
+        tree_call = tree.name == variable
+        children_calls = sum(self._get_nb_requests(child, variable) for child in tree.children)
 
         return tree_call + children_calls
 
     def get_nb_requests(self, variable: str):
         return sum(self._get_nb_requests(tree, variable) for tree in self.trees)
 
-    def key(self, node):
-        name = node['name']
-        period = node['period']
+    def key(self, node: TraceNode) -> str:
+        name = node.name
+        period = node.period
         return f"{name}<{period}>"
 
     def get_flat_trace(self):
         trace = {}
-        for tree in self._trees:
-            trace = {**self._get_flat_trace(tree), **trace}
+        for node in self.browse_trace():
+            trace.update({  # We don't want cache read to overwrite data about the initial calculation. We therefore use a non-overwriting update.
+                key: node_trace
+                for key, node_trace in self._get_flat_trace(node).items()
+                if key not in trace
+                })
         return trace
 
-    def serialize(self, value):
+    def get_serialized_flat_trace(self):
```

Comment:
> Serializing is very expensive for large populations. This PR splits the flattening from the serialization so that we can get one without the other.

Reply:
It's best to go one step further and move these responsibilities out of the Tracer class, just as we did with ComputationLog and PerformanceLog (and as I argued should be done for the new functionality).

The point of `get_serialized_flat_trace` is to perform the work of the `/trace` endpoint. I'm not sure I understand what role this PR gives to the non-serialized version, `get_flat_trace`. But I don't think they belong in the Tracer class.